src/enzyme_ad/jax/Utils.cpp

```diff
@@ -583,6 +583,130 @@ bool canApplyNoNanPattern(bool allowOnFloatingPointMath, Type outTy, Type inTy,
   return allowOnFloatingPointMath || guaranteedNoNanResult(op, rewriter);
 }
 
+bool SymmetricResultAnalysis::constantIntCheck(DenseElementsAttr attr) {
+  return false; // TODO
+}
+
+bool SymmetricResultAnalysis::constantFloatCheck(DenseElementsAttr attr) {
+  for (auto elem : attr.getValues<APFloat>()) {
+    // ?
+  }
+  return false; // TODO
+}
+
+SymmetricResultAnalysis::State 
+SymmetricResultAnalysis::localGuaranteed(Operation *op,
+                                         SmallVectorImpl<Operation *> &localtodo,
+                                         PatternRewriter &rewriter) {
+  assert(op);
+  
+  // TODO will not work right now because constantIntCheck and constantFloatCheck are not implemented
+  if (auto constantOp = dyn_cast<stablehlo::ConstantOp>(op)) {
+    if (guaranteed(constantOp, rewriter)) {
+      return State::GUARANTEED;
+    } else {
+      return State::NOTGUARANTEED;
+    }
+  }
+
+  // A + A^T will always be symmetric
+  if (auto addOp = dyn_cast<stablehlo::AddOp>(op)) {
+    auto lhs = addOp.getLhs();
+    auto rhs = addOp.getRhs();
+
+    if (auto rhsT = dyn_cast_or_null<stablehlo::TransposeOp>(rhs.getDefiningOp())) {
+      auto rhsInput = rhsT.getOperand();
+      if (rhsInput == lhs)
+        return State::GUARANTEED;
+    }
+
+    if (auto lhsT = dyn_cast_or_null<stablehlo::TransposeOp>(lhs.getDefiningOp())) {
+      auto lhsInput = lhsT.getOperand();
+      if (lhsInput == rhs)
+        return State::GUARANTEED;
+    }
+  }
+
+  // A * A^T will always be symmetric
+  if (auto mulOp = dyn_cast<stablehlo::MulOp>(op)) {
```

Comment:
in principle any binary elementwise op with $a$, $a^T$ whould yield a symmetric tensor